File: examples/camera_do_setzoom_individual.cpp
#include "stdio.h"
#include"payloadSdkInterface.h"

// Global pointer to the payload SDK wrapper so callbacks can access it
PayloadSdkInterface* my_payload = nullptr;
void onPayloadStatusChanged(int event, double* param);

bool time_to_exit = false;

#if (CONTROL_METHOD == CONTROL_UART)
T_ConnInfo s_conn = {
    CONTROL_UART,
    payload_uart_port,
    payload_uart_baud
};
#else
T_ConnInfo s_conn = {
    CONTROL_UDP,
    udp_ip_target,
    udp_port_target
};
#endif

void quit_handler(int sig);

int main(int argc, char *argv[]){
	printf("[Init] Starting payload zoom control example...\n");
	signal(SIGINT,quit_handler);

	// Create the payload SDK object with the chosen connection settings
	my_payload = new PayloadSdkInterface(s_conn);

	// Initialise the connection to the payload
	my_payload->sdkInitConnection();
	printf("[Init] Waiting for payload handshake...\n");

	// Subscribe to status updates so we can read acknowledgements and zoom levels
	my_payload->regPayloadStatusChanged(onPayloadStatusChanged);

	// Verify the payload link comes up before issuing commands
	my_payload->checkPayloadConnection();

	// Request periodic zoom level updates (in milliseconds)
	my_payload->setParamRate(PARAM_EO_ZOOM_LEVEL, 1000);
	my_payload->setParamRate(PARAM_IR_ZOOM_LEVEL, 1000);

	// Continuously demonstrate EO/IR zoom transitions until the user exits
	while(1){
		printf("\n[EO] Switching zoom mode to Super Resolution...\n");
		my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_VIDEO_ZOOM_MODE, PAYLOAD_CAMERA_VIDEO_ZOOM_MODE_SUPER_RESOLUTION, PARAM_TYPE_UINT32);
		usleep(4000000); // allow the payload to process the mode change

		printf("[EO] Forcing view source to EO over IR composite...\n");
		my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_VIEW_SRC, PAYLOAD_CAMERA_VIEW_EOIR, PARAM_TYPE_UINT32);

		printf("[EO] Setting Super Resolution zoom to 1x.\n");
		my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_VIDEO_ZOOM_SUPER_RESOLUTION_FACTOR, ZOOM_SUPER_RESOLUTION_1X, PARAM_TYPE_UINT32);
		usleep(4000000);

		printf("[EO] Stepping Super Resolution zoom up to 4x.\n");
		my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_VIDEO_ZOOM_SUPER_RESOLUTION_FACTOR, ZOOM_SUPER_RESOLUTION_4X, PARAM_TYPE_UINT32);
		usleep(4000000);

		printf("[EO] Driving Super Resolution zoom to 30x.\n");
		my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_VIDEO_ZOOM_SUPER_RESOLUTION_FACTOR, ZOOM_SUPER_RESOLUTION_30X, PARAM_TYPE_UINT32);
		usleep(4000000);

		printf("[EO] Switching to Combine zoom mode for extended range...\n");
		my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_VIDEO_ZOOM_MODE, PAYLOAD_CAMERA_VIDEO_ZOOM_MODE_COMBINE, PARAM_TYPE_UINT32);
		usleep(4000000);
		
		printf("[EO] Combine zoom: resetting to 1x.\n");
		my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_VIDEO_ZOOM_COMBINE_FACTOR, ZOOM_COMBINE_1X, PARAM_TYPE_UINT32);
		usleep(4000000);

		printf("[EO] Combine zoom: jumping to 40x.\n");
		my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_VIDEO_ZOOM_COMBINE_FACTOR, ZOOM_COMBINE_40X, PARAM_TYPE_UINT32);
		usleep(4000000);

		printf("[EO] Combine zoom: pushing to 240x.\n");
		my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_VIDEO_ZOOM_COMBINE_FACTOR, ZOOM_COMBINE_240X, PARAM_TYPE_UINT32);
		usleep(4000000);

		printf("[IR] Switching view source to IR over EO composite...\n");
		my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_VIEW_SRC, PAYLOAD_CAMERA_VIEW_IREO, PARAM_TYPE_UINT32);

		printf("[IR] Setting zoom to 1x.\n");
		my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_IR_ZOOM_FACTOR, ZOOM_IR_1X, PARAM_TYPE_UINT32);
		usleep(4000000);

		printf("[IR] Stepping zoom to 4x.\n");
		my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_IR_ZOOM_FACTOR, ZOOM_IR_4X, PARAM_TYPE_UINT32);
		usleep(4000000);

		printf("[IR] Increasing zoom to 8x for maximum magnification.\n");
		my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_IR_ZOOM_FACTOR, ZOOM_IR_8X, PARAM_TYPE_UINT32);
		usleep(4000000);

		usleep(1000); // short delay before repeating the demonstration
	}

	return 0;
}

void quit_handler( int sig ){
    printf("\n[Exit] Received interrupt, shutting down cleanly...\n\n");

    // Close the interface so the payload releases the connection
    try {
        my_payload->sdkQuit();
    }
    catch (int error){}

    // Terminate the sample application
    exit(0);
}

void onPayloadStatusChanged(int event, double* param){
    
    switch(event){
    case PAYLOAD_ACK:{
        printf("[Status] Command %.0f acknowledged with result %.2f\n", param[0], param[1]);
        break;
    }
	case PAYLOAD_PARAM_EXT_ACK:{
		printf("[Status] Extended ACK received: %.2f\n", param[0]);
		break;
	}
	case PAYLOAD_PARAMS:{
		// param[0]: parameter index returned by the payload
		// param[1]: latest value for that parameter
		if(param[0] == PARAM_EO_ZOOM_LEVEL){
			printf("[Status] EO zoom level reported at %.2f\n", param[1]);
		}
		else if(param[0] == PARAM_IR_ZOOM_LEVEL){
			printf("[Status] IR zoom level reported at %.2f\n", param[1]);
		}
		break;
	}
    default: break;
    }
}

File: examples/camera_do_setzoom_individual.py
#!/usr/bin/env python3
import sys
import os

# Add the libs directory to the path
sys.path.append(os.path.join(os.path.dirname(__file__), '..', 'libs'))

# Import config first to setup environment automatically
from config import config

import time
import signal
from pymavlink import mavutil
from payload_sdk import PayloadSdkInterface, payload_status_event_t, payload_param_t
from payload_define import *

my_payload = None
time_to_exit = False

# Handle SIGINT so the example can close cleanly
def quit_handler(sig, frame):
    global my_payload, time_to_exit
    print("\nTERMINATING AT USER REQUEST")
    time_to_exit = True

    # Close payload interface
    try:
        my_payload.sdkQuit()
    except Exception as e:
        print(f"Error while quitting payload: {e}")

    # End program    
    sys.exit(0)

# Callback function for payload status changes
def onPayloadStatusChanged(event, param):
    if payload_status_event_t(event) == payload_status_event_t.PAYLOAD_ACK:
        print(f" --> Got ack, from command: {param[0]:.0f} - result: {param[1]:.2f}")

    elif payload_status_event_t(event) == payload_status_event_t.PAYLOAD_PARAM_EXT_ACK:
        print(f" --> Got ext_ack, result {param[0]:.2f}")

    elif payload_status_event_t(event) == payload_status_event_t.PAYLOAD_PARAMS:
        # param[0]: payload_param_t enum value
        # param[1]: parameter value
        if payload_param_t(param[0]) == payload_param_t.PARAM_EO_ZOOM_LEVEL:  
            print(f"Payload EO_ZOOM_LEVEL: {param[1]:.2f}")

        elif payload_param_t(param[0]) == payload_param_t.PARAM_IR_ZOOM_LEVEL:  
            print(f"Payload IR_ZOOM_LEVEL: {param[1]:.2f}")


def main():
    global my_payload, time_to_exit

    print("Starting ConnectPayload example...\n")
    signal.signal(signal.SIGINT, quit_handler)

    # Create payloadsdk object
    my_payload = PayloadSdkInterface()
    
    # Init payload
    my_payload.sdkInitConnection()
    print("Waiting for payload signal!\n")
    
    # Register callback function
    my_payload.regPayloadStatusChanged(onPayloadStatusChanged)

    # Check connection
    my_payload.checkPayloadConnection()
    
    # Register to receive zoom status updates
    my_payload.setParamRate(payload_param_t.PARAM_EO_ZOOM_LEVEL, 1000)  
    my_payload.setParamRate(payload_param_t.PARAM_IR_ZOOM_LEVEL, 1000)  

    # Perform zoom operations in a loop
    while not time_to_exit:

        # Change EO zoom mode to Super Resolution
        print("\n[EO] Switching zoom mode to Super Resolution...\n")
        my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_VIDEO_ZOOM_MODE, payload_camera_video_zoom_mode.PAYLOAD_CAMERA_VIDEO_ZOOM_MODE_SUPER_RESOLUTION, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
        time.sleep(1)

        # Set view source to EO/IR
        print("[EO] Forcing view source to EO over IR composite...\n")
        my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_VIEW_SRC, payload_camera_view_src.PAYLOAD_CAMERA_VIEW_EOIR, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
        time.sleep(1)  

        # Zoom EO to 1x
        print("[EO] Setting Super Resolution zoom to 1x.\n")
        my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_VIDEO_ZOOM_SUPER_RESOLUTION_FACTOR, payload_camera_video_zoom_super_resolution_factor.ZOOM_SUPER_RESOLUTION_1X, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
        time.sleep(4)

        # Zoom EO to 4x
        print("[EO] Stepping Super Resolution zoom up to 4x.\n")
        my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_VIDEO_ZOOM_SUPER_RESOLUTION_FACTOR, payload_camera_video_zoom_super_resolution_factor.ZOOM_SUPER_RESOLUTION_4X, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
        time.sleep(4)

        # Zoom EO to 30x
        print("[EO] Driving Super Resolution zoom to 30x.\n")
        my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_VIDEO_ZOOM_SUPER_RESOLUTION_FACTOR, payload_camera_video_zoom_super_resolution_factor.ZOOM_SUPER_RESOLUTION_30X, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
        time.sleep(4)

        # Change EO zoom mode to Combine mode
        print("[EO] Switching to Combine zoom mode for extended range...\n")
        my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_VIDEO_ZOOM_MODE, payload_camera_video_zoom_mode.PAYLOAD_CAMERA_VIDEO_ZOOM_MODE_COMBINE, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)

        # Zoom EO to 1x
        print("[EO] Combine zoom: resetting to 1x.\n")
        my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_VIDEO_ZOOM_COMBINE_FACTOR, payload_camera_video_zoom_combine_factor.ZOOM_COMBINE_1X, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
        time.sleep(4)

        # Zoom EO to 40x
        print("[EO] Combine zoom: jumping to 40x.\n")
        my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_VIDEO_ZOOM_COMBINE_FACTOR, payload_camera_video_zoom_combine_factor.ZOOM_COMBINE_40X, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
        time.sleep(6)

        # Zoom EO to 240x
        print("[EO] Combine zoom: pushing to 240x.\n")
        my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_VIDEO_ZOOM_COMBINE_FACTOR, payload_camera_video_zoom_combine_factor.ZOOM_COMBINE_240X, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
        time.sleep(5)

        # Set view source to IR/EO
        print("[IR] Switching view source to IR over EO composite...\n")
        my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_VIEW_SRC, payload_camera_view_src.PAYLOAD_CAMERA_VIEW_IREO, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)

        # Zoom IR to 1x
        print("[IR] Setting zoom to 1x.\n")
        my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_IR_ZOOM_FACTOR, payload_camera_ir_zoom_factor.ZOOM_IR_1X, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
        time.sleep(4)

        # Zoom IR to 4x
        print("[IR] Stepping zoom to 4x.\n")
        my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_IR_ZOOM_FACTOR, payload_camera_ir_zoom_factor.ZOOM_IR_4X, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
        time.sleep(4)

        # Zoom IR to 8x
        print("[IR] Increasing zoom to 8x for maximum magnification.\n")
        my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_IR_ZOOM_FACTOR, payload_camera_ir_zoom_factor.ZOOM_IR_8X, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
        time.sleep(4)

        # Short delay to prevent high CPU usage
        time.sleep(0.001)

if __name__ == "__main__":
    main()
